File: postgre-setup/go/cmd/stress/verify.go
package main

import (
	"flag"
	"fmt"
	"math"
	"math/rand"
	"os"
	"strconv"
	"sync"
	"time"
)

func cmdVerify(args []string) {
	fs := flag.NewFlagSet("verify", flag.ExitOnError)
	apiURL := fs.String("api-url", "http://localhost:8080", "API base URL")
	opsPerAccount := fs.Int("ops-per-account", 5, "Deposits and withdrawals per account for scenario 1")
	transferCount := fs.Int("transfer-count", 500, "Number of random transfers for scenario 2")
	amount := fs.String("amount", "1", "Amount per operation")
	fs.Parse(args)

	client := NewLedgerClient(*apiURL)
	fmt.Println("Ledger correctness tests")
	fmt.Printf("API: %s\n\n", *apiURL)

	fmt.Println("Loading accounts...")
	allAccounts, err := client.GetAllUsers(2000)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error loading accounts: %v\n", err)
		os.Exit(1)
	}
	if len(allAccounts) < 10 {
		fmt.Fprintf(os.Stderr, "Need at least 10 accounts. Got %d. Run: ledger-stress seed --count 1000\n", len(allAccounts))
		os.Exit(1)
	}
	pool := allAccounts
	if len(pool) > 1000 {
		pool = pool[:1000]
	}
	fmt.Printf("Using %d accounts\n\n", len(pool))

	allPass := true

	// Scenario 1
	fmt.Println("━━━ Scenario 1: Deposit + Withdraw Balance Invariant ━━━")
	fmt.Println("  For each account: send N deposits and N withdrawals concurrently.")
	fmt.Println("  Invariant: final_balance = initial_balance (when all ops succeed).\n")
	s1start := time.Now()
	s1pass, s1detail := scenario1(client, pool, *opsPerAccount, *amount)
	fmt.Println(s1detail)
	if s1pass {
		fmt.Printf("  ✓ PASSED  (%.1fs)\n\n", time.Since(s1start).Seconds())
	} else {
		fmt.Printf("  ✗ FAILED  (%.1fs)\n\n", time.Since(s1start).Seconds())
		allPass = false
	}

	// Reload balances for scenario 2
	fmt.Println("Reloading balances for scenario 2...")
	refreshed, err := client.GetAllUsers(2000)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error reloading accounts: %v\n", err)
		os.Exit(1)
	}
	poolIDs := make(map[string]bool, len(pool))
	for _, a := range pool {
		poolIDs[a.ID] = true
	}
	var poolRefreshed []AccountInfo
	for _, a := range refreshed {
		if poolIDs[a.ID] {
			poolRefreshed = append(poolRefreshed, a)
		}
	}

	// Scenario 2
	fmt.Println("━━━ Scenario 2: Transfer Sum Invariant ━━━")
	fmt.Println("  Random transfers between accounts.")
	fmt.Println("  Invariant: sum(all balances) is unchanged before and after.\n")
	s2start := time.Now()
	s2pass, s2detail := scenario2(client, poolRefreshed, *transferCount, *amount)
	fmt.Println(s2detail)
	if s2pass {
		fmt.Printf("  ✓ PASSED  (%.1fs)\n\n", time.Since(s2start).Seconds())
	} else {
		fmt.Printf("  ✗ FAILED  (%.1fs)\n\n", time.Since(s2start).Seconds())
		allPass = false
	}

	// Scenario 3
	fmt.Println("━━━ Scenario 3: Ledger Reconciliation ━━━")
	fmt.Println("  SQL: accounts.balance = SUM(CREDIT entries) − SUM(DEBIT entries)")
	fmt.Println("  Checks every account in the database.\n")
	s3start := time.Now()
	s3pass, s3detail := scenario3(client)
	fmt.Println(s3detail)
	if s3pass {
		fmt.Printf("  ✓ PASSED  (%.1fs)\n\n", time.Since(s3start).Seconds())
	} else {
		fmt.Printf("  ✗ FAILED  (%.1fs)\n\n", time.Since(s3start).Seconds())
		allPass = false
	}

	fmt.Println("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━")
	if allPass {
		fmt.Println("Overall: ✓ ALL PASSED")
		os.Exit(0)
	} else {
		fmt.Println("Overall: ✗ SOME FAILED")
		os.Exit(1)
	}
}

// Decimal arithmetic using integer representation at 6 decimal places.
const decimalScale = int64(1_000_000)

func parseDecimal(s string) int64 {
	f, _ := strconv.ParseFloat(s, 64)
	return int64(math.Round(f * float64(decimalScale)))
}

func formatDecimal(n int64) string {
	neg := n < 0
	if neg {
		n = -n
	}
	whole := n / decimalScale
	frac := n % decimalScale
	s := fmt.Sprintf("%d.%06d", whole, frac)
	if neg {
		s = "-" + s
	}
	return s
}

// scenario1: for each account, run N concurrent deposits and N concurrent
// withdrawals of the same amount. The final balance must equal the initial
// balance adjusted for any ops that succeeded or failed.
func scenario1(client *LedgerClient, accounts []AccountInfo, opsPerAccount int, amount string) (bool, string) {
	n := len(accounts)
	initial := make(map[string]int64, n)
	for _, a := range accounts {
		initial[a.ID] = parseDecimal(a.Balance)
	}
	netOps := make(map[string]int, n)

	type indOp struct {
		typ string
		acc AccountInfo
	}
	var allOps []indOp
	for i := 0; i < opsPerAccount; i++ {
		for _, acc := range accounts {
			allOps = append(allOps, indOp{"deposit", acc})
			allOps = append(allOps, indOp{"withdraw", acc})
		}
	}
	rand.Shuffle(len(allOps), func(i, j int) { allOps[i], allOps[j] = allOps[j], allOps[i] })

	const concurrency = 50
	var mu sync.Mutex
	depositOk, withdrawOk, errors := 0, 0, 0

	for i := 0; i < len(allOps); i += concurrency {
		end := i + concurrency
		if end > len(allOps) {
			end = len(allOps)
		}
		chunk := allOps[i:end]

		type res struct {
			typ     string
			accID   string
			success bool
		}
		results := make([]res, len(chunk))
		var wg sync.WaitGroup
		for k, op := range chunk {
			wg.Add(1)
			k, op := k, op
			go func() {
				defer wg.Done()
				token := fmt.Sprintf("dev_%s", op.acc.UserID)
				var r StressResult
				if op.typ == "deposit" {
					r = client.Deposit(op.acc.ID, amount, token)
				} else {
					r = client.Withdraw(op.acc.ID, amount, token)
				}
				results[k] = res{op.typ, op.acc.ID, r.Success}
			}()
		}
		wg.Wait()

		mu.Lock()
		for _, r := range results {
			if r.success {
				if r.typ == "deposit" {
					netOps[r.accID]++
					depositOk++
				} else {
					netOps[r.accID]--
					withdrawOk++
				}
			} else {
				errors++
			}
		}
		mu.Unlock()
	}

	afterAccs, err := client.GetAllUsers(2000)
	if err != nil {
		return false, fmt.Sprintf("  error fetching final balances: %v", err)
	}
	finalBal := make(map[string]int64, len(afterAccs))
	for _, a := range afterAccs {
		finalBal[a.ID] = parseDecimal(a.Balance)
	}

	amountInt := parseDecimal(amount)
	failures := 0
	var lines []string

	for _, acc := range accounts {
		init := initial[acc.ID]
		net := int64(netOps[acc.ID])
		expected := init + net*amountInt
		actual, exists := finalBal[acc.ID]
		if !exists {
			actual = init
		}
		if expected != actual {
			failures++
			if len(lines) < 10 {
				lines = append(lines, fmt.Sprintf("  ✗ %s... expected=%s actual=%s diff=%s",
					acc.ID[:8], formatDecimal(expected), formatDecimal(actual), formatDecimal(actual-expected)))
			}
		}
	}

	detail := fmt.Sprintf("  accounts=%d  ops/account=%d  amount=$%s\n  deposits_ok=%d  withdrawals_ok=%d  errors=%d\n  balance_failures=%d",
		n, opsPerAccount, amount, depositOk, withdrawOk, errors, failures)
	for _, l := range lines {
		detail += "\n" + l
	}
	return failures == 0, detail
}

// scenario2: run random peer-to-peer transfers between accounts. The total
// sum of all balances must be identical before and after.
func scenario2(client *LedgerClient, accounts []AccountInfo, transferCount int, amount string) (bool, string) {
	n := len(accounts)
	var initialSum int64
	for _, a := range accounts {
		initialSum += parseDecimal(a.Balance)
	}

	type transfer struct{ src, dst AccountInfo }
	transfers := make([]transfer, transferCount)
	for i := range transfers {
		srcIdx := rand.Intn(n)
		dstIdx := rand.Intn(n - 1)
		if dstIdx >= srcIdx {
			dstIdx++
		}
		transfers[i] = transfer{accounts[srcIdx], accounts[dstIdx]}
	}

	const concurrency = 50
	var mu sync.Mutex
	ok, failed := 0, 0

	for i := 0; i < len(transfers); i += concurrency {
		end := i + concurrency
		if end > len(transfers) {
			end = len(transfers)
		}
		chunk := transfers[i:end]
		results := make([]bool, len(chunk))
		var wg sync.WaitGroup
		for k, t := range chunk {
			wg.Add(1)
			k, t := k, t
			go func() {
				defer wg.Done()
				r := client.Transfer(t.src.ID, t.dst.ID, amount, fmt.Sprintf("dev_%s", t.src.UserID))
				results[k] = r.Success
			}()
		}
		wg.Wait()
		mu.Lock()
		for _, success := range results {
			if success {
				ok++
			} else {
				failed++
			}
		}
		mu.Unlock()
	}

	afterAccs, err := client.GetAllUsers(2000)
	if err != nil {
		return false, fmt.Sprintf("  error fetching final balances: %v", err)
	}
	afterMap := make(map[string]int64, len(afterAccs))
	for _, a := range afterAccs {
		afterMap[a.ID] = parseDecimal(a.Balance)
	}
	var finalSum int64
	for _, acc := range accounts {
		if v, exists := afterMap[acc.ID]; exists {
			finalSum += v
		} else {
			finalSum += parseDecimal(acc.Balance)
		}
	}

	diff := finalSum - initialSum
	detail := fmt.Sprintf("  accounts=%d  transfers=%d  amount=$%s\n  transfers_ok=%d  transfers_failed=%d\n  initial_sum=%s  final_sum=%s  diff=%s",
		n, transferCount, amount, ok, failed, formatDecimal(initialSum), formatDecimal(finalSum), formatDecimal(diff))
	return diff == 0, detail
}

// scenario3: server-side SQL reconciliation check — accounts.balance must
// equal SUM(CREDIT entries) − SUM(DEBIT entries) for every account.
func scenario3(client *LedgerClient) (bool, string) {
	rec, err := client.Reconcile(0)
	if err != nil {
		return false, fmt.Sprintf("  error: %v", err)
	}
	detail := fmt.Sprintf("  discrepancies=%d", rec.DiscrepancyCount)
	for i, d := range rec.Discrepancies {
		if i >= 10 {
			break
		}
		idPrefix := d.AccountID
		if len(idPrefix) > 8 {
			idPrefix = idPrefix[:8]
		}
		detail += fmt.Sprintf("\n  ✗ %s... balance=%s ledger=%s diff=%s entries=%d",
			idPrefix, d.AccountBalance, d.LedgerBalance, d.Discrepancy, d.EntryCount)
	}
	return rec.OK, detail
}

File: postgre-setup/ts/cli/src/commands/verify.ts
/**
 * Correctness test suite for the ledger.
 *
 * Three scenarios, each operating over ~1000 accounts:
 *
 * Scenario 1 – Deposit+Withdraw Balance Invariant
 *   For every account in the pool: send N concurrent deposits and N concurrent
 *   withdrawals of the same amount, interleaved (creating real row-level lock
 *   contention). Because the counts are symmetric, a fully-correct ledger must
 *   end up at the starting balance for every account. Any lost update, double
 *   application, or balance/ledger divergence surfaces as a failure here.
 *
 * Scenario 2 – Transfer Sum Invariant
 *   Record the total balance across 1000 accounts. Run many random peer-to-peer
 *   transfers between them. After all transfers settle, re-sum the balances.
 *   The total must be identical — transfers must be zero-sum. A bug in the
 *   double-entry logic (e.g., crediting without debiting) shows up here.
 *
 * Scenario 3 – Ledger Reconciliation
 *   Server-side SQL check: for every account in the DB,
 *     accounts.balance  =  SUM(CREDIT entries) − SUM(DEBIT entries)
 *   This catches any divergence between the running balance and the immutable
 *   audit trail, regardless of which code path created the entries.
 */

import { LedgerClient } from "../lib/api-client.js";

// ─── helpers ────────────────────────────────────────────────────────────────

/** Precise decimal arithmetic using integer cents (6 decimal places). */
const SCALE = 1_000_000n;

function toInt(s: string): bigint {
  // round to 6dp to avoid float noise
  return BigInt(Math.round(parseFloat(s) * 1_000_000));
}

function pad(n: number, w = 10): string {
  return String(n).padStart(w);
}

function fmt(n: bigint): string {
  const neg = n < 0n;
  const abs = neg ? -n : n;
  const whole = abs / SCALE;
  const frac = abs % SCALE;
  return `${neg ? "-" : ""}${whole}.${String(frac).padStart(6, "0")}`;
}

type Account = { id: string; user_id: string; balance: string };

// ─── scenario 1: deposit + withdraw → balance invariant ─────────────────────

/**
 * Sends N deposits and N withdrawals of $amount to every account in `pool`
 * concurrently via individual deposit/withdraw endpoints (not the batch
 * endpoint). Using individual requests gives unambiguous per-op success/failure
 * — each HTTP response directly reflects whether the DB applied the op. There
 * is no batch-level error-propagation to reason about.
 *
 * All ops are interleaved and shuffled before dispatch so that deposits and
 * withdrawals race on the same rows (real row-level lock contention). Some ops
 * may fail with lock-timeout/insufficient-funds; those are accounted for in the
 * expected balance.
 *
 * Correctness check: for each account,
 *   final_balance = initial_balance + (deposits_ok − withdrawals_ok) × amount
 *
 * When all operations succeed (expected with ample initial balance):
 *   final_balance = initial_balance
 */
async function scenario1(
  client: LedgerClient,
  accounts: Account[],
  opsPerAccount: number,
  amount: string,
): Promise<{ pass: boolean; detail: string }> {
  const n = accounts.length;
  const initial = new Map(accounts.map((a) => [a.id, toInt(a.balance)]));

  // Track per-account net expected change (deposits - withdrawals) × amount
  const netOps = new Map<string, number>(accounts.map((a) => [a.id, 0]));

  // Build individual ops for every account, then shuffle to maximise
  // interleaving between deposits and withdrawals on the same rows.
  type IndividualOp = { type: "deposit" | "withdraw"; acc: Account };
  const allOps: IndividualOp[] = [];
  for (let i = 0; i < opsPerAccount; i++) {
    for (const acc of accounts) {
      allOps.push({ type: "deposit", acc });
      allOps.push({ type: "withdraw", acc });
    }
  }
  for (let i = allOps.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [allOps[i], allOps[j]] = [allOps[j], allOps[i]];
  }

  // Dispatch all ops concurrently (max 50 in-flight). Individual endpoints
  // return unambiguous success/failure per-op — no batch poisoning to worry
  // about — so netOps tracking is always accurate.
  const CONCURRENCY = 50;
  let depositOk = 0, withdrawOk = 0, errors = 0;

  for (let i = 0; i < allOps.length; i += CONCURRENCY) {
    const chunk = allOps.slice(i, i + CONCURRENCY);
    const results = await Promise.all(
      chunk.map(({ type, acc }) => {
        const token = `dev_${acc.user_id}`;
        return type === "deposit"
          ? client.deposit(acc.id, amount, token)
          : client.withdraw(acc.id, amount, token);
      })
    );
    for (let k = 0; k < chunk.length; k++) {
      const { type, acc } = chunk[k];
      const r = results[k];
      if (r.success) {
        const delta = type === "deposit" ? 1 : -1;
        netOps.set(acc.id, (netOps.get(acc.id) ?? 0) + delta);
        if (type === "deposit") depositOk++; else withdrawOk++;
      } else {
        errors++;
      }
    }
  }

  // Fetch final balances — use the same large limit as the initial load to
  // avoid non-deterministic ordering dropping pool accounts out of the window.
  const afterResp = await client.getAllUsers(2000);
  const finalBal = new Map(afterResp.map((a) => [a.id, toInt(a.balance)]));

  const amountInt = toInt(amount);
  let failures = 0;
  const lines: string[] = [];

  for (const acc of accounts) {
    const init = initial.get(acc.id)!;
    const net = BigInt(netOps.get(acc.id) ?? 0);
    const expected = init + net * amountInt;
    const actual = finalBal.get(acc.id) ?? init;
    if (expected !== actual) {
      failures++;
      lines.push(
        `  ✗ ${acc.id.slice(0, 8)}... expected=${fmt(expected)} actual=${fmt(actual)} diff=${fmt(actual - expected)}`
      );
    }
  }

  const detail = [
    `  accounts=${n}  ops/account=${opsPerAccount}  amount=$${amount}`,
    `  deposits_ok=${depositOk}  withdrawals_ok=${withdrawOk}  errors=${errors}`,
    `  balance_failures=${failures}`,
    ...lines.slice(0, 10),
  ].join("\n");

  return { pass: failures === 0, detail };
}

// ─── scenario 2: transfer sum invariant ─────────────────────────────────────

/**
 * Executes `transferCount` random peer-to-peer transfers of $amount between
 * accounts in `pool`. All transfers run concurrently (max 50 in-flight).
 *
 * Correctness check: sum of all account balances must be identical before and
 * after. Any bug that creates or destroys money — e.g., crediting without
 * debiting, or debiting twice — changes the sum.
 *
 * Note: some transfers may fail (contention / lock timeout). That is fine
 * because a failed transfer is a no-op: the sum is still preserved. Only a
 * partially-applied transfer (debit without credit, or vice versa) would
 * change the sum.
 */
async function scenario2(
  client: LedgerClient,
  accounts: Account[],
  transferCount: number,
  amount: string,
): Promise<{ pass: boolean; detail: string }> {
  const n = accounts.length;
  const initialSum = accounts.reduce((s, a) => s + toInt(a.balance), 0n);

  // Generate random transfer pairs (src ≠ dst)
  type Transfer = { src: Account; dst: Account };
  const transfers: Transfer[] = [];
  for (let i = 0; i < transferCount; i++) {
    let srcIdx = Math.floor(Math.random() * n);
    let dstIdx = Math.floor(Math.random() * (n - 1));
    if (dstIdx >= srcIdx) dstIdx++;
    transfers.push({ src: accounts[srcIdx], dst: accounts[dstIdx] });
  }

  // Execute transfers concurrently (max 50 in-flight)
  const CONCURRENCY = 50;
  let ok = 0, failed = 0;

  for (let i = 0; i < transfers.length; i += CONCURRENCY) {
    const chunk = transfers.slice(i, i + CONCURRENCY);
    const results = await Promise.all(
      chunk.map(({ src, dst }) =>
        client.transfer(src.id, dst.id, amount, `dev_${src.user_id}`)
      )
    );
    for (const r of results) {
      if (r.success) ok++; else failed++;
    }
  }

  // Fetch final balances and re-sum
  const afterResp = await client.getAllUsers(2000);
  const afterMap = new Map(afterResp.map((a) => [a.id, toInt(a.balance)]));
  const finalSum = accounts.reduce((s, a) => s + (afterMap.get(a.id) ?? toInt(a.balance)), 0n);

  const diff = finalSum - initialSum;
  const pass = diff === 0n;

  const detail = [
    `  accounts=${n}  transfers=${transferCount}  amount=$${amount}`,
    `  transfers_ok=${ok}  transfers_failed=${failed}`,
    `  initial_sum=${fmt(initialSum)}  final_sum=${fmt(finalSum)}  diff=${fmt(diff)}`,
  ].join("\n");

  return { pass, detail };
}

// ─── scenario 3: ledger reconciliation ──────────────────────────────────────

async function scenario3(
  client: LedgerClient,
): Promise<{ pass: boolean; detail: string }> {
  const rec = await client.reconcile();
  const lines: string[] = [
    `  discrepancies=${rec.discrepancy_count}`,
  ];
  for (const d of rec.discrepancies.slice(0, 10)) {
    lines.push(
      `  ✗ ${d.account_id.slice(0, 8)}... balance=${d.account_balance} ledger=${d.ledger_balance} diff=${d.discrepancy} entries=${d.entry_count}`
    );
  }
  return { pass: rec.ok, detail: lines.join("\n") };
}

// ─── main ────────────────────────────────────────────────────────────────────

export async function verifyCommand(opts: {
  apiUrl: string;
  opsPerAccount: number;
  transferCount: number;
  amount: string;
}) {
  const client = new LedgerClient(opts.apiUrl);

  console.log(`Ledger correctness tests`);
  console.log(`API: ${opts.apiUrl}\n`);

  // Load the full account pool (up to 2000)
  console.log("Loading accounts...");
  const allAccounts = await client.getAllUsers(2000);
  if (allAccounts.length < 10) {
    console.error(`Need at least 10 accounts. Got ${allAccounts.length}. Run: ledger-stress seed --count 1000`);
    process.exit(1);
  }
  // Use up to 1000 accounts for each scenario
  const pool = allAccounts.slice(0, Math.min(1000, allAccounts.length));
  console.log(`Using ${pool.length} accounts\n`);

  let allPass = true;

  // ── Scenario 1 ──────────────────────────────────────────────────────────
  console.log("━━━ Scenario 1: Deposit + Withdraw Balance Invariant ━━━");
  console.log("  For each account: send N deposits and N withdrawals concurrently.");
  console.log("  Invariant: final_balance = initial_balance (when all ops succeed).\n");
  const s1start = performance.now();
  const s1 = await scenario1(client, pool, opts.opsPerAccount, opts.amount);
  const s1elapsed = ((performance.now() - s1start) / 1000).toFixed(1);
  console.log(s1.detail);
  console.log(`  ${s1.pass ? "✓ PASSED" : "✗ FAILED"}  (${s1elapsed}s)\n`);
  if (!s1.pass) allPass = false;

  // Reload balances after scenario 1 for scenario 2
  console.log("Reloading balances for scenario 2...");
  const refreshed = await client.getAllUsers(2000);
  const poolRefreshed = refreshed.filter((a) => pool.some((p) => p.id === a.id));

  // ── Scenario 2 ──────────────────────────────────────────────────────────
  console.log("━━━ Scenario 2: Transfer Sum Invariant ━━━");
  console.log("  Random transfers between accounts.");
  console.log("  Invariant: sum(all balances) is unchanged before and after.\n");
  const s2start = performance.now();
  const s2 = await scenario2(client, poolRefreshed, opts.transferCount, opts.amount);
  const s2elapsed = ((performance.now() - s2start) / 1000).toFixed(1);
  console.log(s2.detail);
  console.log(`  ${s2.pass ? "✓ PASSED" : "✗ FAILED"}  (${s2elapsed}s)\n`);
  if (!s2.pass) allPass = false;

  // ── Scenario 3 ──────────────────────────────────────────────────────────
  console.log("━━━ Scenario 3: Ledger Reconciliation ━━━");
  console.log("  SQL: accounts.balance = SUM(CREDIT entries) − SUM(DEBIT entries)");
  console.log("  Checks every account in the database.\n");
  const s3start = performance.now();
  const s3 = await scenario3(client);
  const s3elapsed = ((performance.now() - s3start) / 1000).toFixed(1);
  console.log(s3.detail);
  console.log(`  ${s3.pass ? "✓ PASSED" : "✗ FAILED"}  (${s3elapsed}s)\n`);
  if (!s3.pass) allPass = false;

  // ── Summary ─────────────────────────────────────────────────────────────
  console.log("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━");
  console.log(`Overall: ${allPass ? "✓ ALL PASSED" : "✗ SOME FAILED"}`);
  process.exit(allPass ? 0 : 1);
}
